src/openfermion/transforms/_code_transform_functions_test.py

```diff
@@ -0,0 +1,93 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import unittest
+
+from openfermion.transforms._code_transform_functions import *
+from openfermion.hamiltonians import MolecularData
+from openfermion.transforms import binary_code_transform
+from openfermion.transforms import get_fermion_operator
+from openfermion.utils import eigenspectrum
+
+
+def lih_hamiltonian():
+    geometry = [('Li', (0., 0., 0.)), ('H', (0., 0., 1.45))]
+    active_space_start = 1
+    active_space_stop = 3
+    molecule = MolecularData(geometry, 'sto-3g', 1,
+                             description="1.45")
+    molecule.load()
+    molecular_hamiltonian = molecule.get_molecular_hamiltonian(
+        occupied_indices=range(active_space_start),
+        active_indices=range(active_space_start, active_space_stop))
+    hamiltonian = get_fermion_operator(molecular_hamiltonian)
+    ground_state_energy = eigenspectrum(hamiltonian)[0]
+    return hamiltonian, ground_state_energy
+
+
+class CodeTransformTest(unittest.TestCase):
+    def test_checksum_code(self):
+        hamiltonian, gs_energy = lih_hamiltonian()
+        code = checksum_code(4, 0)
+        qubit_hamiltonian = binary_code_transform(hamiltonian, code)
+        self.assertAlmostEqual(gs_energy,
+                               eigenspectrum(qubit_hamiltonian)[0])
+
+    def test_jordan_wigner(self):
+        hamiltonian, gs_energy = lih_hamiltonian()
+        code = jordan_wigner_code(4)
+        qubit_hamiltonian = binary_code_transform(hamiltonian, code)
```

Comment:
Is this supposed to do the same thing as the already-implemented `jordan_wigner`? If so, can we add a test that it does? Same for Bravyi-Kitaev.

Reply:
@conta877 @kevinsung @babbush I wonder if it would make sense to have `jordan_wigner` alias to this (eliminating duplicate code) - does being more general mean that this code is slower / is there any other reason not to?

Reply:
I agree with @kevinsung 's suggestion but am very opposed to @idk3 's suggestion. The implementation of Jordan-Wigner using the CodeOperator is significantly more complex. The Jordan-Wigner code might be useful to forks of OpenFermion that don't involve CodeOperator. The performance is also likely better in the more simple implementation.

Reply:
I agree that we should keep both. I also think its a good example of code transforms for people. I didn't check jordan-wigner implementation but the Bravyi-Kitaevs produce different qubitOperators.

Reply:
Really ? Let me have a look at that - I tend to accidentally  switch the two matrices. In any case it seems to be a valid transform, since it passed the test.

Reply:
A good test would be to see if the transformations are all isospectral. How different is the output of the transformation? If there is a broad family of Bravyi-Kitaev like transformations with truly different structures it would be quite interesting.

Reply:
I just tested it the code transform vs. bravyi_kitaev, and frankly I am a bit confused.  $a^\dagger_i a_i$ should return a projector .5*(1-Z_F(i)), right ? so F(6) is empty, and I find the following:
>>> bravyi_kitaev(FermionOperator('6^ 6'))
(0.5+0j) [] +
(-0.5+0j) [Z3 Z5 Z6]
>>> binary_code_transform(FermionOperator('6^ 6'),bravyi_kitaev_code(19))
0.5 [] +
-0.5 [Z6]

Reply:
Our BK transform is implemented as in arXiv:1701.07072.

Reply:
Eq. 5 is what I mean (sorry it is  {F(i) U i}, not just F(i)). So it seems the flip sets are different.

Reply:
If you run
```
bravyi_kitaev(FermionOperator('6^ 6'), n_qubits=8)
```
the output is
```
(0.5+0j) [] +
(-0.5+0j) [Z6]
```

Reply:
indeed. if you set n_qubits=14, it says:
(0.5+0j) [] +
(-0.5+0j) [Z3 Z5 Z6]

Why is it qubit dependent ?

Reply:
Because the mode with the highest index is always the root of the Fenwick tree (see arXiv:1701.07072).

Reply:
Implementing our version, I followed the instructions in 1208.5986 and S. Tranter's paper. There  they segment the matrix if its in between two dimension that are consecutive integer powers of two. It seems like the Fenwick-tree mapping is different.

Reply:
@msteudtner I think you're right. Perhaps this PR should use our implementation of Fenwick trees (contained in `transforms`) to define the Bravyi-Kitaev code? @babbush do you have an opinion? It seems that the code definitions are fundamentally different between 1208.5986 and 1701.07072.

Reply:
Ok I just saw your comment below.

Reply:
Based on the discussion below, I'm content with just checking that the JW (not BK) gives the same QubitOperators as the existing implementation, and that the new BK gives isospectral operators (does not need to give identical QubitOperators).